### PR TITLE
Feature: add header image overlay setting

### PR DESCRIPTION
**Pull request by JoshuaHungDinh**
Resolves [GIVE-396]

## Description
This pull request introduces an overlay setting that can be applied to the header image of the Visual Form Builder. The setting is accessible from the Styles tab in the Design Settings sidebar and is applicable to all form templates and selected image styles. With this setting, users can select a color from the WordPress palette and adjust the opacity to their preference.

Since the HeaderImage is an image tag, we can not reliably create an overlay with a pseudo element. To achieve the overlay an additional div was included in the HeaderImage template. The settings for both the color & opacity are then passed to the overlay as CSS variables.

## Affects
Visual Form Builder
Header Image

## Visuals

https://github.com/impress-org/givewp/assets/75056371/7efd1f4c-8983-4e8d-a70c-0d2cbed827c3

## Testing Instructions

- Create a form with the VFB & upload a headerImage.
- Select the "Style" tab and click the overlay setting to set a color.
- After a color is chosen you should see the opacity setting. 
- Using the range control slider adjust that the opacity changes accordingly.
- Using the number input update the opacity again.
- Test with all header image styles "background, center, above" & with all available templates.

## Pre-review Checklist
-   [x] Acceptance criteria satisfied and marked in related issue
-   [ ] Relevant `@unreleased` tags included in DocBlocks
-   [ ] Includes unit tests
-   [x] Reviewed by the designer (if follows a design)
-   [x] [Self Review](https://give.gitbook.io/development-manual/devops/github/code-reviews#self-review) of code and UX completed

[GIVE-396]: https://stellarwp.atlassian.net/browse/GIVE-396?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: src/DonationForms/ViewModels/DonationFormViewModel.php [hunk 1]
```diff
@@ -212,26 +213,26 @@ public function exports(): array
         $formDesign = $this->getFormDesign($this->designId());
 
         return [
-            'donateUrl' => $donateUrl,
-            'validateUrl' => $validateUrl,
-            'authUrl' => $authUrl,
+            'donateUrl'            => $donateUrl,
```
Comment (jonwaldstein) [inline src/DonationForms/ViewModels/DonationFormViewModel.php:216]: something weird going on with the formatting here

File: src/DonationForms/Properties/FormSettings.php [hunk 2]
```diff
@@ -311,6 +323,10 @@ public static function fromArray(array $array): self
             $array['designSettingsTextFieldStyle']
         ) : DesignSettingsTextFieldStyle::DEFAULT();
 
+        $self->designSettingsImageColor = $array['designSettingsImageColor'];
+
+        $self->designSettingsImageOpacity = $array['designSettingsImageOpacity'];
```
Comment (jonwaldstein) [inline src/DonationForms/Properties/FormSettings.php:328]: We'll probably want to do a nullish conditional on these to fallback to when there is no image color or opacity

File: src/DonationForms/resources/registrars/templates/layouts/HeaderImage.tsx [hunk 3]
```diff
@@ -3,6 +3,17 @@ import {HeaderImageProps} from '@givewp/forms/propTypes';
 /**
  * @since 3.5.0
  */
-export default function HeaderImage({url, alt}: HeaderImageProps) {
-    return <img src={url} alt={alt} />;
+export default function HeaderImage({url, alt, color, opacity}: HeaderImageProps) {
+    // @ts-ignore
```
Comment (jonwaldstein) [inline src/DonationForms/resources/registrars/templates/layouts/HeaderImage.tsx:7]: Can we try to appease typescript without the ignore?

File: src/FormBuilder/resources/js/form-builder/src/settings/design/style-controls/header-image/index.tsx [hunk 4]
```diff
@@ -0,0 +1,71 @@
+import {BaseControl, PanelBody, RangeControl} from '@wordpress/components';
+import {__} from '@wordpress/i18n';
+import {setFormSettings, useFormState} from '@givewp/form-builder/stores/form-state';
+import {PanelColorSettings, SETTINGS_DEFAULTS} from '@wordpress/block-editor';
+import useDonationFormPubSub from '@givewp/forms/app/utilities/useDonationFormPubSub';
+
+export default function HeaderImage({dispatch}) {
```
Comment (jonwaldstein) [inline src/FormBuilder/resources/js/form-builder/src/settings/design/style-controls/header-image/index.tsx:7]: Please add docblock with unreleased

File: src/FormBuilder/resources/js/form-builder/src/settings/design/style-controls/index.tsx [hunk 5]
```diff
@@ -13,12 +13,14 @@ const getDesign = (designId: string) => formDesigns[designId];
...
     const {
-        settings: {designId},
+        settings: {designSettingsImageUrl},
     } = useFormState();
+    const dispatch = useFormStateDispatch();
 
-    const isClassicTemplate = !getDesign(designId).isMultiStep;
+    const {
+        settings: {},
+    } = useFormState();
```
Comment (jonwaldstein) [inline src/FormBuilder/resources/js/form-builder/src/settings/design/style-controls/index.tsx:23]: This looks like it can be removed.

**Review by jonwaldstein — CHANGES_REQUESTED** [review]
Looking good! just noticed a few minor things

**Comment by jonwaldstein** [comment]
@JoshuaHungDinh 2 more things:

1. There's a failing unit test we'll need to address from updating the form settings
2. I was expecting the overlay settings to be next to the header image and then realized they were in the styles tab lol.  I'm honestly not sure which is better but wanted to bring that up for discussion cc: @jdghinson

File: src/DonationForms/resources/registrars/templates/layouts/HeaderImage.tsx (hunk 3 above)
Comment (JoshuaHungDinh) [reply src/DonationForms/resources/registrars/templates/layouts/HeaderImage.tsx]: Yes, this was actually resolved by adding the React.CSSProperties. I removed the @ts-ignore flag. 
Resolved: da19cf635bc405ee85b11b5919e26dc0d9c8721e

File: src/FormBuilder/resources/js/form-builder/src/settings/design/style-controls/header-image/index.tsx (hunk 4 above)
Comment (JoshuaHungDinh) [reply src/FormBuilder/resources/js/form-builder/src/settings/design/style-controls/header-image/index.tsx:10]: Resolved: f02623f0f22cc1a6628ae9b7536863962766a319

File: src/FormBuilder/resources/js/form-builder/src/settings/design/style-controls/index.tsx (hunk 5 above)
Comment (JoshuaHungDinh) [reply src/FormBuilder/resources/js/form-builder/src/settings/design/style-controls/index.tsx]: Resolved: 0437d2cc1238c1f709ac9c9bee3b82d46efcef39

File: src/DonationForms/ViewModels/DonationFormViewModel.php (hunk 1 above)
Comment (JoshuaHungDinh) [reply src/DonationForms/ViewModels/DonationFormViewModel.php:216]: Screen Shot 2024-03-08 at 9 53 42 AM

I have arrows aligned as the default format in my ide.

File: src/DonationForms/Properties/FormSettings.php (hunk 2 above)
Comment (JoshuaHungDinh) [reply src/DonationForms/Properties/FormSettings.php]: Resolved: b37aa50cafcf0ebf2552a8a66d8dabce295a7c08

File: src/DonationForms/ViewModels/DonationFormViewModel.php (hunk 1 above)
Comment (jonwaldstein) [reply src/DonationForms/ViewModels/DonationFormViewModel.php:216]: @JoshuaHungDinh yes please

Comment (JoshuaHungDinh) [reply src/DonationForms/ViewModels/DonationFormViewModel.php]: Resolved: dc395febde39970bb5a8639e710ec5e2b6611e00

**Review by jonwaldstein — APPROVED** [review]
Great job man!

**Pull request merged**